### ISSUE TITLE
Unable to use `__kconfig`

### ISSUE THREAD

**Issue by eiffel-fl** [cilium/ebpf#698]
**Describe the bug**
With `cilium/ebpf`, it is not possible reference kconfig variables like `LINUX_KERNEL_VERSION`.

**To Reproduce**
Steps to reproduce the behavior. Please include:

```sh
# Try to bpf2go a file which contains the following line:
# extern int LINUX_KERNEL_VERSION __kconfig;
foo_bpfel.go: can't load BPF from ELF: load BTF: reference to .kconfig: not supported
```

**Expected behavior**
We should be able to reference kconfig variables.

**Comment by lmb** [comment]
At a very high level:
1. You need a parser for kconfig style files.
2. You need to extract the Datasec patching logic from `CollectionSpec.RewriteConstants` into a separate function.
3. You need to modify `elfCode.loadDataSections` to add a special case for `.kconfig` sections. In that special case you will have to iterate the variables in the section and patch the correct values from (1) into the raw `data`.
4. You need to remove the block on `.kconfig` from `fixupDatasec`

**Comment by ti-mo** [comment]
@eiffel-fl This is `internal.KernelVersion()`.

**Comment by lmb** [comment]
Yes, that's the idea. You can probably resurrect [patchValues](https://github.com/cilium/ebpf/commit/1ff15d21014336f653b09fe3842977f01c5945d2#diff-5ed7da7003773d8e0a4dca00d4fc08fc5582695eefe22a400db15edfd5aab840L1285-L1338) for step 2.

> I think I will give a try to implement the whole __kconfig parsing.

I think Timo has a point, can we land this in two steps? First LINUX_VERSION_CODE, followed by full kconfig support later?

**Comment by lmb** [comment]
From https://facebookmicrosites.github.io/bpf/blog/2020/02/19/bpf-portability-and-co-re.html#dealing-with-kernel-version-and-configuration-differences

```C
extern u32 LINUX_KERNEL_VERSION __kconfig;
extern u32 CONFIG_HZ __kconfig;
```

`__kconfig` is from https://elixir.bootlin.com/linux/v5.19.1/source/tools/lib/bpf/bpf_helpers.h#L174:

```C
#define __kconfig __attribute__((section(".kconfig")))
```

So there should be a `.kconfig` section with a single VAR that has extern linkage. It seems like that isn't the case for you?
 Probably the first thing that needs to be fixed.

> Then another problem that I met is that the BPF loader complains about extern linkage:

libbpf probably replaces extern with global linkage, as you suspect. Maybe you can dig up the code that does it?

If yes, you'd have to extend BTF marshaling like we do for [StripFuncLinkage](https://github.com/cilium/ebpf/blob/7bbc7384cbe1fec7c7e281786cdb789b5e36f21d/btf/btf.go#L469).

> I am not sure to fully understand, but it seems [libbpf people](https://lore.kernel.org/bpf/20191117070807.251360-7-andriin@fb.com/T/) use a .extern map to trick this

> So far, I wrote a bit of code to add a LINUX_KERNEL_VERSION to .rodata map.

I think it'll be easier if you don't try to synthesise a new `.extern` map and instead use the `.kconfig` generated by the compiler. Let's not stick it in `.rodata` either.

> Also, does cilium/ebpf support extern? I saw some ExternVar/Func in the code but nothing more.

What are you referring to with `extern`? extern linkage in BTF? extern variables in BPF code?

> eBPF code tries to load address 0x0:

This is probably because the dword load into `r1` isn't tagged with the correct map name, and hence isn't rewritten to point at the correct map.

**Comment by lmb** [comment]
> I do not know what you mean by "section" but the `.kconfig` is not printed by either `readelf` or `objdump`.

Okay, I played with this myself: https://github.com/lmb/ebpf/commit/0fd50437c2e02d625dd1cf9acc2e0314576a2498 Indeed, there is no `.kconfig` section. Maybe something to do with `extern`.

> I think I will need to write such a function to deal with extern and get their offset from BTF, then replacing the value of the corresponding offset.

Yes, I think you're right. However, I'm not sure I understand why libbpf iterates symbols though instead of directly querying the BTF for `.kconfig` or `.ksym` sections. Maybe you could try a different approach first:

1. Try to find a `btf.Datasec` called `.kconfig`. If it exists, create a map in `CollectionSpec.Maps[".kconfig"]`
2. Figure out when a relocation points at a kconfig section, and make sure WithReference(".kconfig") is called as well as setting the correct offset.

Some complications:
- We should refuse WEAK symbols for now, since we don't support them: https://github.com/libbpf/libbpf/blob/a0d1e22c7790f22809d33b3c5c1e3ded89157cc8/src/libbpf.c#L3713
- We don't really support calculating alignment at the moment: https://github.com/libbpf/libbpf/blob/a0d1e22c7790f22809d33b3c5c1e3ded89157cc8/src/libbpf.c#L3738 Maybe the `int` support we have is enough though? https://github.com/cilium/ebpf/blob/aaa9af24fdd8e397be823526928782cc93d59188/btf/types.go#L678-L687

**Comment by ti-mo** [comment]
It's fairly straightforward, the only issue I have is that if we parse /proc/config into a map[string]string, it consumes a non-negligible amount of memory that will remain cached for the lifetime of the application.

I'd like us to experiment with a kind of associative array that doesn't store the full key, e.g. map[uint64]string, and only supports one-way lookups. I'd be interested to hear if it's worth it.

**Comment by ti-mo** [comment]
> What do you mean by the "lifetime of the application"?

Many long-running apps embed ebpf-go, so any package globals we populate will remain in memory until the app terminates.

> Cannot we just drop the map[string]string once we finished dealing with the __kconfig "section"?

That would cause multiple ELF loads with kconfig entries to re-parse /proc/config every time, which is rather expensive.

Scratch my comment about the one-way map; it's just an optimization that can be done later.

> I will open a PR to add this feature to what we have now.

The point I was trying to make: before you open any PRs, please open a thread in #ebpf-go-dev on Slack with your proposed design so we can give early feedback. Ideally, we want to review a PR at most twice and have most of the technical discussion out of the way before the PR is created. This causes the least amount of friction for everyone involved.

Feel free to take code from https://github.com/ti-mo/kconfig to get started, so you don't have to write a parser from scratch. It's old code and needs some optimization, but it does the job. I'll probably remove that package once it has been lifted into the lib.
